MISC-239 - Initial Development

mattdrees opened the pull request:
This adds the initial Cru checkstyle rules, as well as the derived rules for Wsapi.

For https://jira.cru.org/browse/MISC-239.

mattdrees committed:
remove junit

mattdrees committed:
remove unused classes

mattdrees committed:
add description and url

mattdrees committed:
add a simple .gitignore

mattdrees committed:
add a simple .gitattributes

mattdrees committed:
define cru_checks checkstyle config using google_checks

This uses the 'google_checks.xml' checkstyle ruleset that is distributed with checkstyle.
Since checkstyle doesn't yet support a native configuration override mechanism [1],
we'll use an RFC 5261 xml diff utility so that we can add and subtract from google_checks.xml
without copying and maintaining the whole file.

[1]: https://github.com/checkstyle/checkstyle/issues/2873

mattdrees committed:
define wsapi_checks in terms of cru_checks

Like the previous commit, we'll use xml diffs
to maintain (legacy) project-specific variations from cru_checks.

Note: I thought about putting this in wsapi itself,
but this approach is less complex.
I'm not yet totally convinced this is the better approach,
but I think it's the best way to start.

mattdrees committed:
add scm config

mattdrees committed:
add a basic readme

mattdrees committed:
require whitespace after commas

mattdrees committed:
strengthen requirement on newlines & operators

Instead of just dropping Google's requirement, we'll require the opposite.
Cru's codebases don't use the `nl`, and there's not a good reason to switch.

mattdrees commented:
@wrandall22 @frett I converted this into a PR. That's what I should have done in the first place.

mattdrees committed:
use error instead of warning

It's not clear why checkstyle decided to use 'warning':
https://github.com/checkstyle/checkstyle/issues/4452

wrandall22 commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+# ignore maven target directory
+target/
```
What about IntelliJ files and things like `.DS_STORE`?

mattdrees replied:
Yeah, good idea.

I have a global gitignore that ignores those, but probably worth putting them in the project gitignore.

wrandall22 reviewed (commented):
One comment

mattdrees committed:
ignore intellij and DS_Store files

mattdrees committed:
relax line length limit to 120

mattdrees committed:
header doc

mattdrees committed:
switch to RFC 7351 format

I like the namespace to help differentiate patch elements from checkstyle elements.

mattdrees committed:
enforce short methods

mattdrees committed:
wsapi line lengths are fixed now

mattdrees committed:
update comments; these will be short-term

frett reviewed (approved):
Looks good to me

mattdrees commented:
thanks guys!
